[CDAP-18694] Create artifact localizer sidecar for preview runner when enabled in cConf

wyzhang opened the pull request:
Why:
Preview runner may run arbitrary user code, so in some deployment
env it may be refrained from accessing artifact jar directly
(e.g. CDAP deployed in GKE and artifacts stored in GCS)

What:
This change adds an option to allow starting up artifact localizer
sidecar container, which can fetch and cache artifact on behalf
this preview runner (to be implemented in next PR)

masoud-io commented on cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/preview/DistributedPreviewManager.java:
```diff
@@ -157,22 +160,45 @@ public void run() {
+                .build());
+          }
+
+          LOG.info("Starting preview runners with {} instances", runnerResourceSpec.getInstances());
```
nit: add whether sidecard container is also being used or not to the log.

wyzhang replied:
Done

masoud-io commented on cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/preview/PreviewRunnerTwillApplication.java:
```diff
@@ -32,24 +34,37 @@
-      .build();
+      .apply();
+
+    if (artifactLocalizerResourceSpec.isPresent()) {
```
nit: use ifPresent (and consumer) to avoid if block.

wyzhang replied:
Done

masoud-io commented on cdap-common/src/main/resources/cdap-default.xml:
```diff
@@ -3240,6 +3240,18 @@
+    <value>false</value>
+    <description>
+      Whether to enable artifact localizer sidecar container to fetch and cache artifacts.
+      This replaces the default direct access of artifact jars by preview runners.
```
nit:
if true, artifacts are accessed through side-car container instead of directly accessing them through GCS.

wyzhang replied:
Done

masoud-io reviewed (approved):
LGTM. nits are optional.

wyzhang committed:
[CDAP-18694] Create artifact localizer sidecar for preview runner
              when enabled in cConf

Why:
Preview runner may run arbitrary user code, so in some deployment
env it may be refrained from accessing artifact jar directly
(e.g. CDAP deployed in GKE and artifacts stored in GCS)

What:
This change adds an option to allow starting up artifact localizer
sidecar container, which can fetch and cache artifact on behalf
this preview runner (to be implemented in next PR)
